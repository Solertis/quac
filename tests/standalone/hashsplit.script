#!/bin/bash

# Copyright (c) 2012-2013 Los Alamos National Security, LLC, and others.

. ./environment.sh

cd $DATADIR

set +e
exec 2>&1  # send stderr to stdout to make output more readable


# Check argument parsing

x hashsplit
echo $?
x hashsplit 2
echo $?
x hashsplit foo
echo $?
x hashsplit 0 foo
echo $?
y "hashsplit '' foo"
echo $?
y "hashsplit 2 ''"
echo $?


# Check failure on can't open output file(s)

y "hashsplit 2 /does/not/exist < /dev/null"
echo $?


# Check that it works (note that we repeat "a" several times to make sure
# we're hashing just the key and not the whole line), including with UTF-8.

cat > in.txt <<EOF
a	1
b	2
c	3
a	4
b	5
a	6
a	7
a	8
a	9
a	10
a	11
a	12
a	13
a	14
a	15
a	16
a	17
a	18
a	19
a	20
a	21
nullvaluewithtab	
nullvaluenotab
私の名前は中野です
EOF

x cat in.txt
y "hashsplit 2 out < in.txt"
x cat out.0
x cat out.1

# Now split with much wider degree. This is mainly used to test that other
# hash functions are consistent. The value 240 is from the default open files
# limit of 256 (ulimit -n) on OS X Mountain Lion.
#
# Note that both "nullvaluenotab" and the Japanese text hash mod 240 is 19. I
# am pretty well convinced this is a coincidence. I ran /usr/share/dict/words
# through hashsplit and ended up with approximately 17% maximum variation in
# the size of 100 buckets, which seems reasonable. The actual hash values of
# the two strings differ.
y "hashsplit 240 out < in.txt"
x cat out.7
x cat out.19
