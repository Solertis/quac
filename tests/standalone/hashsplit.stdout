$ hashsplit
usage: hashsplit N BASENAME

Split standard input containing a stream of key/value lines separated
by a single tab into N output files named BASENAME.i according to the
hash values of the keys. The value may be absent, either with or
without a tab following the key. Keys and values may contain any bytes
except zero, tab, and newline.
1
$ hashsplit 2
usage: hashsplit N BASENAME

Split standard input containing a stream of key/value lines separated
by a single tab into N output files named BASENAME.i according to the
hash values of the keys. The value may be absent, either with or
without a tab following the key. Keys and values may contain any bytes
except zero, tab, and newline.
1
$ hashsplit foo
usage: hashsplit N BASENAME

Split standard input containing a stream of key/value lines separated
by a single tab into N output files named BASENAME.i according to the
hash values of the keys. The value may be absent, either with or
without a tab following the key. Keys and values may contain any bytes
except zero, tab, and newline.
1
$ hashsplit 0 foo
invalid number of output files: 0
1
$ (hashsplit '' foo)
invalid number of output files: 0
1
$ (hashsplit 2 '')
length of BASENAME cannot be 0
1
$ (hashsplit 2 /does/not/exist < /dev/null)
can't open /does/not/exist.0: No such file or directory
1
$ cat in.txt
a	1
b	2
c	3
a	4
b	5
a	6
a	7
a	8
a	9
a	10
a	11
a	12
a	13
a	14
a	15
a	16
a	17
a	18
a	19
a	20
a	21
nullvaluewithtab	
nullvaluenotab
私の名前は中野です
$ (hashsplit 2 out < in.txt)
$ cat out.0
a	1
c	3
a	4
a	6
a	7
a	8
a	9
a	10
a	11
a	12
a	13
a	14
a	15
a	16
a	17
a	18
a	19
a	20
a	21
nullvaluewithtab	
$ cat out.1
b	2
b	5
nullvaluenotab
私の名前は中野です
$ (hashsplit 240 out < in.txt)
$ cat out.7
b	2
b	5
$ cat out.19
nullvaluenotab
私の名前は中野です
